File: go/109.go
package main

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func build(arr []int, l, r int) *TreeNode {
  if l > r {
    return nil
  }
  var mid int = l + (r-l)/2
  return &TreeNode{
    Val: arr[mid],
    Left: build(arr, l, mid-1),
    Right: build(arr, mid+1, r),
  }
}

func sortedListToBST(head *ListNode) *TreeNode {
	if head == nil {
		return nil
	}
	var arr []int = make([]int, 0)
	for n := head ; n != nil ; n=n.Next {
	  arr = append(arr, n.Val)
	}
	return build(arr, 0, len(arr)-1)
}

File: cpp/109.cpp
#include <iostream>
#include <vector>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  TreeNode *build(vector<int> &arr, int l, int r) {
    if (l > r)
      return nullptr;
    int mid = l + (r - l) / 2;
    TreeNode *root = new TreeNode(arr[mid]);
    root->left = build(arr, l, mid - 1);
    root->right = build(arr, mid + 1, r);
    return root;
  }
  TreeNode *sortedListToBST(ListNode *head) {
    if (!head)
      return nullptr;
    vector<int> arr;
    for (ListNode *cur = head; cur; cur = cur->next)
      arr.push_back(cur->val);
    int size = arr.size();
    return build(arr, 0, size - 1);
  }
};
